aten/src/ATen/native/cuda/layer_norm_kernel.cu

```diff
@@ -0,0 +1,408 @@
+    T* Y) {
+  const int64_t i = blockIdx.x;
+  for (int64_t j = threadIdx.x; j < N; j += blockDim.x) {
+    const int64_t index = i * N + j;
+    const T gamma_v = gamma == nullptr ? T(1) : gamma[j];
+    const T beta_v = beta == nullptr ? T(0) : beta[j];
+    Y[index] = (X[index] - mean[i]) * rstd[i] * gamma_v + beta_v;
+  }
+}
+
+template <typename T>
+__global__ void ComputeInternalGradientsCUDAKernel(
+    int64_t N,
+    const T* dY,
+    const T* X,
+    const T* gamma,
+    T* ds,
+    T* db) {
+  using T_ACC = acc_type<T, true>;
+  const int64_t i = blockIdx.x;
+  T_ACC sum1 = 0;
+  T_ACC sum2 = 0;
+  for (int j = threadIdx.x; j < N; j += blockDim.x) {
+    const int64_t index = i * N + j;
+    const T gamma_v = gamma == nullptr ? T(1) : gamma[j];
+    sum1 += dY[index] * X[index] * gamma_v;
+    sum2 += dY[index] * gamma_v;
+  }
+  sum1 = BlockReduceSum<T_ACC>(sum1);
+  sum2 = BlockReduceSum<T_ACC>(sum2);
+  if (threadIdx.x == 0) {
+    ds[i] = sum1;
+    db[i] = sum2;
+  }
+}
+
+template <typename T>
+__global__ void ComputeGradientFusedParamsCUDAKernel(
+    int64_t M,
+    int64_t N,
+    const T* mean,
+    const T* rstd,
+    const T* ds,
+    const T* db,
+    T* c1,
+    T* c2) {
+  const int64_t index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (index < M) {
+    const T s = T(1) / static_cast<T>(N);
+    const T a = (db[index] * mean[index] - ds[index]) * rstd[index] *
+        rstd[index] * rstd[index] * s;
+    c1[index] = a;
+    c2[index] = -(a * mean[index] + db[index] * rstd[index] * s);
+  }
+}
+
+template <typename T>
+__global__ void LayerNormBackwardCUDAKenrel(
+    int64_t N,
+    const T* dY,
+    const T* X,
+    const T* gamma,
+    const T* a,
+    const T* b,
+    const T* c,
+    T* dX) {
+  const int64_t i = blockIdx.x;
+  for (int64_t j = threadIdx.x; j < N; j += blockDim.x) {
+    const int64_t index = i * N + j;
+    const T gamma_v = gamma == nullptr ? T(1) : gamma[j];
+    dX[index] = a[i] * dY[index] * gamma_v + b[i] * X[index] + c[i];
+  }
+}
+
+template <typename T>
+__global__ void GammaBetaBackwardSimpleCUDAKernel(
+    int64_t M,
+    int64_t N,
+    const T* dY,
+    const T* X,
+    const T* mean,
+    const T* rstd,
+    T* dg,
+    T* db) {
+  const int64_t j = blockIdx.x * blockDim.x + threadIdx.x;
+  if (j < N) {
+    T sum1 = 0;
+    T sum2 = 0;
+    for (int64_t i = 0; i < M; ++i) {
+      const int64_t index = i * N + j;
+      sum1 += dg == nullptr ? T(0) : dY[index] * (X[index] - mean[i]) * rstd[i];
+      sum2 += db == nullptr ? T(0) : dY[index];
+    }
+    if (dg != nullptr) {
+      dg[j] = sum1;
+    }
+    if (db != nullptr) {
+      db[j] = sum2;
+    }
+  }
+}
+
+template <typename T>
+__global__ void GammaBetaBackwardCUDAKernel(
+    int64_t M,
+    int64_t N,
+    const T* dY,
+    const T* X,
+    const T* mean,
+    const T* rstd,
+    T* dg,
+    T* db) {
+  using T_ACC = acc_type<T, true>;
+  __shared__ T_ACC g_shared[C10_WARP_SIZE][C10_WARP_SIZE + 1];
```

Comment:
Block dimensions are hardcoded below, and shared memory size is defined through an unrelated macro that happens to be correct. This is fragile.

Reply:
Done